ledger-core/v2/virtual/execute/execute_test.go

```diff
@@ -156,14 +156,104 @@ func TestSMExecute_StartRequestProcessing(t *testing.T) {
 
-	{ // update known requests panics
+	{
+		// call same request
+		// expect duplicate detection and SM stop
+		wasStoped := false
+
 		execCtx := smachine.NewExecutionContextMock(mc).
-			UseSharedMock.Set(shareddata.CallSharedDataAccessor)
+			UseSharedMock.Set(shareddata.CallSharedDataAccessor).
+			LogMock.Return(smachine.Logger{}).
+			StopMock.Set(
+			func() (s1 smachine.StateUpdate) {
+				wasStoped = true
+				return smachine.StateUpdate{}
+			})
 
-		checkerFunc := func() {
-			smExecute.stepStartRequestProcessing(execCtx)
-		}
-		assert.Panics(t, checkerFunc, "panic with not implemented deduplication algorithm should be here")
+		smExecute.stepStartRequestProcessing(execCtx)
+		assert.Equal(t, true, wasStoped)
+	}
+
+	mc.Finish()
+}
+
+func TestSMExecute_Deduplication(t *testing.T) {
+	var (
+		ctx = inslogger.TestContext(t)
+		mc  = minimock.NewController(t)
+
+		pd              = pulse.NewFirstPulsarData(10, longbits.Bits256{})
+		pulseSlot       = conveyor.NewPresentPulseSlot(nil, pd.AsRange())
+		smObjectID      = gen.UniqueIDWithPulse(pd.PulseNumber)
+		smGlobalRef     = reference.NewSelf(smObjectID)
+		smObject        = object.NewStateMachineObject(smGlobalRef)
+		sharedStateData = smachine.NewUnboundSharedData(&smObject.SharedState)
+
+		callFlags = payload.BuildCallFlags(contract.CallIntolerable, contract.CallDirty)
+	)
+
+	smObjectAccessor := object.SharedStateAccessor{SharedDataLink: sharedStateData}
+	request := &payload.VCallRequest{
+		Polymorph:           uint32(payload.TypeVCallRequest),
+		CallType:            payload.CTConstructor,
+		CallFlags:           callFlags,
+		CallSiteDeclaration: testwallet.GetPrototype(),
+		CallSiteMethod:      "New",
+		CallOutgoing:        smObjectID,
+		Arguments:           insolar.MustSerialize([]interface{}{}),
+	}
+
+	smExecute := SMExecute{
+		Payload:           request,
+		pulseSlot:         &pulseSlot,
+		objectSharedState: smObjectAccessor,
+	}
+
+	smExecute = expectedInitState(ctx, smExecute)
+
+	{
+		// duplicate pending request exists
+		// expect SM stop
+		pendingList := smObject.PendingTable.GetList(contract.CallIntolerable)
+		pendingList.Add(smExecute.execution.Outgoing)
+
+		wasStoped := false
+
+		execCtx := smachine.NewExecutionContextMock(mc).
+			UseSharedMock.Set(shareddata.CallSharedDataAccessor).
+			LogMock.Return(smachine.Logger{}).
+			StopMock.Set(
+			func() (s1 smachine.StateUpdate) {
+				wasStoped = true
+				return smachine.StateUpdate{}
+			})
+
+		smExecute.stepDeduplication(execCtx)
+		require.Equal(t, true, wasStoped)
+	}
+
+	{
+		// start deduplication before getting all pending requests
+		// expecting going sleep
+		smObject.PendingTable = object.NewRequestTable()
+
+		execCtx := smachine.NewExecutionContextMock(mc).
+			UseSharedMock.Set(shareddata.CallSharedDataAccessor).
+			AcquireForThisStepMock.Return(false).
+			SleepMock.Return(smachine.NewStateConditionalBuilderMock(mc).ThenRepeatMock.Return(smachine.StateUpdate{}))
```

Comment:
Do not inline mock in mock code, pls :) It degrades readability of code

Reply:
done